edb/schema/pointers.py

```diff
@@ -322,8 +359,16 @@ def merge_targets(cls, schema, ptr, t1, t2, *, allow_contravariant=False):
 
             return schema, current_target
 
-    def get_derived(self, schema, source, target, *,
-                    derived_name_base=None, **kwargs):
+    def get_derived(
+        self,
+        schema: s_schema.Schema,
+        source: so.Object,
```

Comment:
```suggestion
        source: sources.Source,
```

Reply:
I initially annotated this way. But then I had several issues in other methods, especially `get_or_create_union_pointer`, for example due to this variable swap:

```python
    if direction is PointerDirection.Inbound:
        source, target = target, source
```

I opted for keeping the source annotated as a more high-level object for this reason. Should I keep the `source` parameter in `get_or_create_union_pointer` as `so.Object`, and then assert that source is an instance of `sources.Source` before calling `genptr.get_derived` method?

Reply:
The `source` parameter to  `get_or_create_union_pointer` should be `Source` as well.

Reply:
Ok, in this case I propose to add a `type: ignore` in the line where they are swapped. Because MyPy complains properly there, that:

```
Incompatible types in assignment (expression has type "Type", variable has type "Source")
Incompatible types in assignment (expression has type "Source", variable has type "Type")
```